secret_store/src/key_server.rs

```diff
@@ -14,111 +14,204 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
+use std::thread;
+use std::sync::Arc;
+use std::sync::mpsc;
+use futures::{self, Future};
+use parking_lot::Mutex;
+use tokio_core::reactor::Core;
 use ethcrypto;
 use ethkey;
 use super::acl_storage::AclStorage;
 use super::key_storage::KeyStorage;
+use key_server_cluster::ClusterCore;
 use traits::KeyServer;
-use types::all::{Error, RequestSignature, DocumentAddress, DocumentEncryptedKey};
+use types::all::{Error, RequestSignature, DocumentAddress, DocumentEncryptedKey, ClusterConfiguration};
+use key_server_cluster::{ClusterClient, ClusterConfiguration as NetClusterConfiguration};
 
 /// Secret store key server implementation
-pub struct KeyServerImpl<T: AclStorage, U: KeyStorage> {
-	acl_storage: T,
-	key_storage: U,
+pub struct KeyServerImpl {
+	data: Arc<Mutex<KeyServerCore>>,
 }
 
-impl<T, U> KeyServerImpl<T, U> where T: AclStorage, U: KeyStorage {
+/// Secret store key server data.
+pub struct KeyServerCore {
+	close: Option<futures::Complete<()>>,
+	_handle: Option<thread::JoinHandle<()>>,
```

Comment:
The [JoinHandle](https://doc.rust-lang.org/std/thread/struct.JoinHandle.html) doesn't join the thread on `Drop`. If you want that behavior, you'll have to trigger it manually. If not, then it's not necessary to keep around.